File: 46. Permutations/Dynamic/permutations.rb
def number_from_list (list)
	out = 0
	len_list = list.length()
	base = 10
	
	for i in 0...len_list
		out += (list[i] * (base ** (len_list - 1 - i)))
	end
	
	return out
end

def number_to_array (number, original_list_len = 0)
	out = []
	base = 10
	running_number = number
	
	while running_number > 0
		modulo = running_number % base
		running_number -= modulo
		running_number /= base
		out << (modulo)
	end
	
	if original_list_len != 0
		while (out.length < original_list_len)
			out << 0
		end
	end
	
	out = out.reverse()
	
	return out
end

def list_comparison (list1, list2)
	out = false
	len_list1 = list1.length
	len_list2 = list2.length
	
	if (len_list1 != len_list2)
		return out
	end
	
	list1_bool_map = []
	list2_bool_map = []
	
	for i in 0...len_list1
		list1_bool_map << false
		list2_bool_map << false
	end
	
	for i in 0...len_list1
		for j in 0...len_list2
			if (list1_bool_map[i] == false) and (list2_bool_map[j] == false)
				if (list1[i] == list2[j])
					list1_bool_map[i] = true
					list2_bool_map[j] = true
					break
				end
			end
		end
	end
	
	bool_counter = 0
	for i in 0...len_list1
		if (list1_bool_map[i] and list2_bool_map[i])
			bool_counter += 1
		end
	end
	
	return (list1_bool_map.length == bool_counter)
end

def appears (element, list)
	out = -1
	
	for i in 0...list.length
		if (element == list[i])
			out += 1
		end
	end
	
	return out
end

nums = [1,2,3]
len_nums = nums.length
nums = nums.sort()
lowest_value_nums_number = number_from_list(nums)
nums = nums.reverse()
highest_value_nums_number = number_from_list(nums)

puts(lowest_value_nums_number)
puts(highest_value_nums_number)

permutations = []

running_number = lowest_value_nums_number

while (running_number <= highest_value_nums_number)
	temp_nums_list = number_to_array(running_number, len_nums)
	if (list_comparison(temp_nums_list, nums))
		permutations << temp_nums_list
	end
	
	running_number += 1
end

puts(String(permutations))

File: 46. Permutations/Dynamic/permutations.php
<?php
function numberFromList ($list){
	$out = 0;
	$lenList = count($list);
	$base = 10;
	
	for ($i = 0; $i < $lenList; $i++){
		$out += ($list[$i] * pow($base,($lenList - (1 + $i))));
	}
	
	return $out;
}

function numberToList ($number, $originalListLen = 0){
	$out = [];
	$base = 10;
	$running_number = $number;
	
	while ($running_number > 0){
		$modulo = $running_number % $base;
		$running_number -= $modulo;
		$running_number /= $base;
		array_push($out, $modulo);
	}
	
	if ($originalListLen != 0){
		while (count($out) < $originalListLen){
			array_push($out, 0);
		}
	}
	
	$out = array_reverse($out);
	
	return ($out);
}

function listComparison ($list1, $list2){
	$out = false;
	$lenList1 = count($list1);
	$lenList2 = count($list2);
	
	if ($lenList1 != $lenList2){
		return ($out);
	}
	
	$list1BoolMap = [];
	$list2BoolMap = [];
	
	for ($i = 0; $i < $lenList1; $i++){
		array_push($list1BoolMap, false);
		array_push($list2BoolMap, false);
	}
	
	for ($i = 0; $i < $lenList1; $i++){
		for ($j = 0; $j < $lenList1; $j++){
			if (($list1BoolMap[$i] == false) && ($list2BoolMap[$j] == false)){
				if ($list1[$i] == $list2[$j]){
					$list1BoolMap[$i] = true;
					$list2BoolMap[$j] = true;
					break;
				}
			}
		}
	}
	
	$boolCounter = 0;
	
	for ($i = 0; $i < $lenList1; $i++){
		if ($list1BoolMap[$i] && $list2BoolMap[$i]){
			$boolCounter += 1;
		}
	}
	
	return (count($list1BoolMap) == $boolCounter);
}

function appears ($element, $list){
	$out = -1;
	$lenList = count($list);
	
	for ($i = 0; $i < $lenList; $i++){
		if ($element == $list[$i]){
			$out += 1;
		}
	}
	
	return ($out);
}

$nums = [1,2,3];
$lenNums = count($nums);
sort($nums);
$lowestValueNumsNumber = numberFromList($nums);
$nums = array_reverse($nums);
$highestValueNumsNumber = numberFromList($nums);

echo ($lowestValueNumsNumber . "\n");
echo ($highestValueNumsNumber . "\n");

$permutations = [];

$runningNumber = $lowestValueNumsNumber;

while ($runningNumber <= $highestValueNumsNumber){
	$tempNumsList = numberToList($runningNumber, $lenNums);
	if (listComparison($tempNumsList, $nums)){
		array_push($permutations, $tempNumsList);
	}
	
	$runningNumber++;
}

echo(json_encode($permutations));

echo("\n\n\n");

?>
